### ISSUE TITLE
Report suspicious iota usages

### ISSUE THREAD

**Issue by quasilyte** [go-critic/go-critic#1469]
This code may seem OK, but it contains an error:
```go
const (
	maxCommandID = 9999
	
	ValueGrayFlag = iota + (maxCommandID + 1)
	ValueBlueFlag
	ValueTealFlag
)
```
`ValueGrayFlag` will be 10001 instead of the implied 10000.

In order to fix it, the unrelated const needs to be moved away (or +1 should be removed, but it's sloppy).
```go
const maxCommandID = 9999

const (
	ValueGrayFlag = iota + (maxCommandID + 1)
	ValueBlueFlag
	ValueTealFlag
)
```

Therefore, we can assume that this is a bug pattern. There are legitimate usages of intermixed iota and other constants, but I would predict that we should at least try to find the false positive/negative ratio. Enabling this checker as an experimental addition to go-critic should be good enough.

This linter should report any constants that appear in `const ()` spec where `iota` is also used. All preceding iota constants should be reported (there should be 0).

We should be careful to emphasize on a fact that moving the const away is not always a full solution - iota constants should be revisited (their values might change).

**Comment by ccoVeille** [comment]
Wow, I wouldn't have expected this.

**Comment by cmcpasserby** [comment]
iota is a index for everything in the const block, this makes sense to me since the `maxCommandID` would have ate index 0.

**Comment by ccoVeille** [comment]
Comment from @theckman on Slack Gopher when I shared this oddity after you created the issue

> I guess since I’ve read the spec front to back at least 5 times now I am also not surprised, like Chris. 
> 
> But also, what you shared above isn’t always an error. In the past I’ve used the properties of iota to skip a value I didn’t care about, and it looked similar to the “bad” snippet above. 
> Source: https://gophers.slack.com/archives/C029RQSEE/p1742481752471199?thread_ts=1742479292.573629&cid=C029RQSEE

Also, @sudomateo and @cmcpasserby also discussed about it on Slack Gopher, so I'm quoting them here

**Comment by cristaloleg** [comment]
> I didn’t care about, and it looked similar to the “bad” snippet above.

Tend to disagree. In the original (however, quite rare) case it's not for skipping `iota` value.

**Comment by quasilyte** [comment]
If there is a high chance of misinterpretation, I consider it to be a dangerous thing to use. Just like in C++ there are many edge cases that are usually avoided just to be safe even though you could use the same RTFM argument in there.
If you don't like the diagnostic, turn it off.

**Comment by quasilyte** [comment]
Sir, stop making assumption about my experience and regarding what I do and don't know.

We usually create tickets like this as ideas to share. I personally believe that it could be a potential bug in the code and we should report it. As you can see, I'm not the only one in this camp, so it's not right-vs-wrong but more about the two perspectives.

**Comment by sudomateo** [comment]
> PS: iota isn't an index, it's simply a const value that starts from 0 and gets auto incremented on every other line.

Let's not say that `iota` isn't an index. As already noted in https://github.com/go-critic/go-critic/issues/1469#issuecomment-2740650146, the Go language specification uses the word "index" to describe `iota`. That verbiage and the examples that immediately follow are extremely helpful in understanding how `iota` works.

> Also, RTFM is a somewhat unfriendly tone. You don't have to go that far just to disagree with some idea.

@fredmenez could have better spelled out their intent rather than subjecting their comment to a single character difference but it seems that @quasilyte misread this as "read the fucking manual" (RTFM) instead of "read the manual first" (RTMF). @fredmenez clarified this in a subsequent comment.

> Sir, stop making assumption about my experience and regarding what I do and don't know.
>
> I don't like your condescending tone, I'll make that point again. I can repeat once more, this is about the potential bug, not the "real" bug -- this is why it's a linter, it has to make assumptions sometimes to catch more bugs as opposed to less.

I don't see any evidence of a condescending tone here @quasilyte. I'm observing that the Go community has a habit of jumping to emotion when a discussion or debate is started that challenges their views on a topic. Quite frankly, I'm tired of that behavior. Even after @fredmenez clarified their intent with RTMF you still believed they are being condescending. Why, because they asked you to rephrase your problem statement on issue? Come on, that's a valid ask.

> We usually create tickets like this as ideas to share. I personally believe that it could be a potential bug in the code and we should report it. As you can see, I'm not the only one in this camp, so it's not right-vs-wrong but more about the two perspectives.
>
> Again, it's not about what it is, it's about how people use it, and what people usually expect from it. There is 0 folks that know all features of an arbitrary language. Even Go isn't that small if you consider all the pitfalls you could find.

When tickets like this are created it invites discussion. Discussion will bring opinions that differ from your own and those opinions should be... discussed. You say it's not about right versus wrong but you're trying to convince the thread that your opinion is the right one by denying the discussion points from everyone else. After reading the initial post and the discussion here I'm inclined to agree with @fredmenez that the original post should be updated to better clarify what is meant by "bug pattern" and note that the `iota` specification defines the behavior well.

That being said, there's clearly an issue with people not understanding the quirks of `iota` and perhaps an extra linter rule could help remind people to consult the Go language specification which does a fantastic job of defining and demonstrating the behavior of `iota`. I'm personally not in favor of adding such a linter rule but I can see that it could be valuable to others. I'm also someone that likes to refer to the language specification so that I know exactly what my tools are doing. I encourage other people to do the same.

**Comment by fredmenez** [comment]
Wow i wasn't expecting such a detailed coverage. I must admit I did use unnecessary emoticons initially, which probably obviously didn't help. And then i realised i was driving the whole discussion around defending the understanding of iota rather than safeguarding users of the language which is the intent of this project. The whole misunderstanding is probably due to fatigue, and precision of one intents. Ah well it's easy to argue, see you in the pub to cheer up :)

**Comment by quasilyte** [comment]
@sudomateo it's not my job to be polite to every stranger that comes to the project that is dear to my heart (and was created by me) that does not try to investigate how issues are posted here. You haven't done your homework, so I can only remind you of one; you won't be reminded of that if you're not actively using words that imply that an author of the issue don't know how things work (even if I don't, it's not a good intent to express it like that, we're talking about the code, not about what "I want").

This part in particular:

> If you want ValueGrayFlag to be equal to 10000, a more idiomatic readable fix is this

It's like I came with my own problem and want to solve it. This is a linter project, we discuss potential issues here. And, unlike some other people, I do collect stats.

![Image](https://github.com/user-attachments/assets/d8c38a05-64dc-46a2-8c4d-8fec5dc339ad)

If 50% people could make this mistake in a very separated, clearly provocative form ("there is a trick to it, since it's a quiz"), there are ~x2 more chances to people to make that misinterpretation in the wild, when they have much more stuff in their context.

So, in short, I don't mind spending my time explaining stuff here, even if people don't care. But when I see that it's a random guy that sounds like they put themself in "I know more than you do" it does trigger me. It would be better for me to completely ignore that, it's true. But then again, I have to clarify things right? Like showing you the stats (%), elaborating on the fact that this is a more-or-less heuristics-based project and lining in general is a false positive vs false negative thing, not a simple yes/no topic.

If you come to someone else home (this repo), behave like a polite guest and you will be welcome. Otherwise, don't be surprised that people may not like you (they don't have to, just like you don't have to like me).

**Comment by quasilyte** [comment]
TL;DR: For me, 50% of potential mistakes in a separate quiz is a lot. Even if it was around 25%, I would still consider this checker to be useful. This is what we do - we help people to find a bug and then they can go and read "why it was behaving like that"?

**Comment by quasilyte** [comment]
>  the original post should be updated to better clarify what is meant by "bug pattern" and note that the iota specification defines the behavior well.

I created this post mostly as a self-reminder. I don't have to persuade you to add a checker to this linter. If you need some details, you have them above, 50% of people get this thing wrong.

**Comment by quasilyte** [comment]
You know what? I am mad now.

If something is spelled in the docs/spec, it doesn't matter. You can claim that everything is described in the specs, or can be inferred from that. With this argument, do we treat all bugs as irrelevant? Since you can just go and read the specs.

And this is not true, people make silly bugs all the time. It's a % to make a mistake while using some part of the language. So no, I'm not giving up.

**Comment by sudomateo** [comment]
> @sudomateo it's not my job to be polite to every stranger that comes to the project that is dear to my heart (and was created by me) that does not try to investigate how issues are posted here.
>
> If you come to someone else home (this repo), behave like a polite guest and you will be welcome. Otherwise, don't be surprised that people may not like you (they don't have to, just like you don't have to like me).

Based on the wording here I believe you are conflating feedback about the issue/project with feedback about you as a person. No one here is being condescending to you or diminishing your value as a person. However, people are asking for clarification, rightfully so, on the issue because the initial wording of the issue read as if there's a bug with `iota` which there isn't.

The home analogy is a bit incorrect. We were invited here as guests by nature of this being an open source issue on an open source project. None of the guests here are being impolite. People seem to like this project otherwise they wouldn't be here discussing things. I'd encourage you to try and separate feedback about the project from feedback about you as a person.

> TL;DR: For me, 50% of potential mistakes in a separate quiz is a lot. Even if it was around 25%, I would still consider this checker to be useful. This is what we do - we help people to find a bug and then they can go and read "why it was behaving like that?"

Again, you're using the word "bug" here when really the issue is that careless usage of `iota` can result in undesired behavior.

50% is a lot! Thank you for taking the time to make the quiz and post the results. I think that's a good indicator that a linter rule could provide value to people I even though I, Matthew Sanabria, do not personally find value in the linter for my use case.

> If something is spelled in the docs/spec, it doesn't matter. You can claim that everything is described in the specs, or can be inferred from that. With this argument, do we treat all bugs as irrelevant? Since you can just go and read the specs.

It does matter, but not for the reason you may think. Having something defined in a specification is helpful for when the community disagrees on something. The specification can then be used to clarify behavior and help the community reach consensus. It's not some weapon to be wielded, but rather a reference to show that, no, there is no bug with `iota` itself and, yes, there is a lack of clarity around its usage that can justify a linter rule.

> I didn't foresee that this idea would result in this. I kinda regret I came to visit this project again. I stopped caring a long time ago.
>
> You know what? I am mad now.

Respectfully, this is childish behavior. Now I am commenting on you as a person. There was nothing but gratitude and thanks for your work on this project. If this project really is your "home" then why would you regret coming to an issue where your users are gathering to discuss the issue because they actually like and use your project? Why would you be mad about that?

People would love to have an open source project that people cared enough about to comment on with kindness. I'm unsure if you're burned out or otherwise exhausted and defensive but I'd recommend taking a moment to reflect and see that no one here is being unkind to you.

**Comment by quasilyte** [comment]
> However, people are asking for clarification, rightfully so, on the issue because the initial wording of the issue read as if there's a bug with iota which there isn't.

Because for whatever reason I need to clarify something and persuade with numbers while for a random comment maker it's enough to write "it's described in the specs". So no, I don't believe it's "rightfully so". Even a small research could be a good contribution. An investigation or a poll could help here, it doesn't have to be me who conducts it.

You can't just go and ask someone to do it for you. If you have doubts - you can check it yourself.

**Comment by quasilyte** [comment]
We even had a tool like https://www.quasilyte.dev/gocorpus/ to make stats collection easier. Numbers speak louder than words, especially words that start with "<x> is wrong". Instead of that, there are countless ways to express the concern and add something valuable to the discussion.

The reason why many open source maintainers are burned out is partially this. You invest a lot of time, you need to respect all of your users, but they won't spend time trying to understand the project/issue they're commenting, and all of this for free.

**Comment by quasilyte** [comment]
> We were invited here as guests by nature of this being an open source issue on an open source project.

This is one way of interpreting it, but I have a different opinion. You can freely use the project, but if you want to be welcome in the discussion that happens between people (as opposed to abstract beings), it's better to start in a different way. Of course you can jump in a random project and demand stuff, or claim that some idea (that wasn't interpreted properly) is wrong, but I don't think that it's productive. Remember, nobody invited a particular comment for a discussion.

If it was a different comment, we wouldn't be in this situation.

**Comment by cristaloleg** [comment]
Ok, it's time to stop with the discussion not related to the original case. Any further comment not about checker will be removed and author will be banned.

Thanks.

**Comment by sudomateo** [comment]
> I would, but in the meanwhile consider the asymmetrical nature of this issue as well. I had to spend more time on this issue because somebody was not convinced. There is a contribution (research/stats/examples), and there is a subjective comment in any for of "I do disagree". The latter does not help the project in any way.

Considered! Both objective and subjective comments are important, especially for a linter which, by its very nature, contains subjective rules. One of which we're discussing right now.

> The reason why many open source maintainers are burned out is partially this. You invest a lot of time, you need to respect all of your users, but they won't spend time trying to understand the project/issue they're commenting, and all of this for free.

I'd consider the time I'm spending discussing this with you to be time spent understanding the issue, the project, and even you as a person.

> Subjective opinions are overrated. Not commenting on something is sometimes better than leaving the comment. No value was gained here (but the cost was higher than zero).

This advice goes both ways of course. There is cost, in time, from anyone reading and writing comments, not just the maintainers. I simply agreed with feedback about the initial wording of the issue and gave my opinion in whether the linter rule would be beneficial.

> In a sense, you were never invited here. You decided to come here yourself.

I was mentioned on this issue specifically. I consider that an invite. However, now I'm taking your comment to mean that I'm no longer invited here. Is that correct?

**Comment by sudomateo** [comment]
> Ok, it's time to stop with the discussion not related to the original case. Any further comment not about checker will be removed and author will be banned.
> 
> Thanks.

Oops! Sorry I didn't see this load when I was already typing my other comment. My apologies!
